### ISSUE TITLE
How to statically link against `libpfring`?

### ISSUE THREAD

**Issue by mertyildiran** [ntop/PF_RING#881]
Installed `libpfring` using:

```shell
apt-get update && \
    apt-get -y -q install wget lsb-release && \
    wget -q http://apt.ntop.org/20.04/all/apt-ntop.deb && dpkg -i apt-ntop.deb && \
    apt-get clean all && \
    apt-get update && \
    apt-get -y install pfring
```

Considering [this small program](https://github.com/google/gopacket/blob/master/examples/pfdump/main.go) from `gopacket` library.  The build **succeeds** if I dynamically link:

```shell
go build -o main .
```

But if I statically link:

```shell
go build -ldflags="-extldflags=-static -s -w" -o main .
```

The build **fails** with:

```text
pfring_utils.c:(.text+0x1ad3): undefined reference to `dlopen'
/usr/bin/ld: pfring_utils.c:(.text+0x1b0b): undefined reference to `dlsym'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_mod_af_xdp.o): in function `pfring_mod_af_xdp_stats':
pfring_mod_af_xdp.c:(.text+0xee): undefined reference to `xsk_socket__fd'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_mod_af_xdp.o): in function `pfring_mod_af_xdp_close':
pfring_mod_af_xdp.c:(.text+0x368): undefined reference to `xsk_socket__delete'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x371): undefined reference to `xsk_umem__delete'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x38b): undefined reference to `bpf_get_link_xdp_id'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x3a1): undefined reference to `bpf_set_link_xdp_fd'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_mod_af_xdp.o): in function `pfring_mod_af_flush_tx_q':
pfring_mod_af_xdp.c:(.text+0x54d): undefined reference to `xsk_socket__fd'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_mod_af_xdp.o): in function `pfring_mod_af_xdp_open':
pfring_mod_af_xdp.c:(.text+0x1302): undefined reference to `xsk_umem__create'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x136a): undefined reference to `xsk_socket__create'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x1568): undefined reference to `xsk_socket__delete'
/usr/bin/ld: pfring_mod_af_xdp.c:(.text+0x1601): undefined reference to `xsk_socket__fd'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_mod_af_xdp.o): in function `pfring_mod_af_xdp_get_selectable_fd':
pfring_mod_af_xdp.c:(.text+0x4d): undefined reference to `xsk_socket__fd'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_ft_zmq.o): in function `zmq_lib_init':
pfring_ft_zmq.c:(.text+0x21): undefined reference to `dlopen'
/usr/bin/ld: pfring_ft_zmq.c:(.text+0x64): undefined reference to `dlsym'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib/libpfring.a(pfring_ft_dpi.o): in function `ndpi_lib_init':
pfring_ft_dpi.c:(.text+0x1d): undefined reference to `dlopen'
/usr/bin/ld: pfring_ft_dpi.c:(.text+0x5c): undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
```

Even though the Go language involved in the example that I provide, my question is not specific to Go language. Since it's FFI-ed to C.

In conclusion, what should I do to statically link against `libpfring`? or is it even possible?

**Comment by cardigliano** [comment]
Hi @mertyildiran, even the examples under https://github.com/ntop/PF_RING/tree/dev/userland/examples are using static linking (you can take a look at the Makefile), note that you need to link -ldl as some pf_ring module is using dlopen to use external libraries.

**Comment by mertyildiran** [comment]
@cardigliano I'm sorry that my question is not clear enough. The static build complains for the lines that calls methods like `dlopen` and `dlsym` which are related to dynamic linking:

https://github.com/ntop/PF_RING/blob/d799e34813da7ef2ffbeb323cede0bbcdbc3828e/userland/lib/pfring_utils.c#L938

https://github.com/ntop/PF_RING/blob/d799e34813da7ef2ffbeb323cede0bbcdbc3828e/userland/lib/pfring_utils.c#L946

When a Go program is built with `-static` flag, it generates a static binary that's not going to be dynamically linked to anything else. Is it possible to disable these lines with a macro or something else?

The methods with `xsk_` prefix also causes `undefined reference` errors which I don't understand the reason. Is it possible to build without `AF_XDP` support? I guess `-D HAVE_PF_RING_ZC` macro should be removed from the `gcc` command. But I'm not sure how I can achieve that through [this line](https://github.com/google/gopacket/blob/master/pfring/pfring.go#L12).

Also does the `AF_XDP` support provide TX packets (outgoing) or is it only RX packets (incoming)?

**Comment by cardigliano** [comment]
@mertyildiran you can use ./configure --disable-xdp to disable the af_xdp support, it will skip the compilation of the pfring_mod_af_xdp.c file all together:

~/PF_RING/userland$ ./configure --disable-xdp

AF_XDP support includes both RX and TX.

**Comment by mertyildiran** [comment]
@cardigliano OK, thank you so much!
